File: backend/springboot/src/main/java/tn/maiko26/springboot/service/ProjectService.java
package tn.maiko26.springboot.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.maiko26.springboot.exception.CustomException;
import tn.maiko26.springboot.model.Project;
import tn.maiko26.springboot.model.Team;
import tn.maiko26.springboot.model.User;
import tn.maiko26.springboot.model.relations.ProjectMember;
import tn.maiko26.springboot.repository.ProjectRepository;
import tn.maiko26.springboot.repository.ProjectMemberRepository;
import tn.maiko26.springboot.repository.UserRepository;

import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;
import java.util.Optional;

@Service
public class ProjectService {

    @Autowired
    private ProjectRepository projectRepository;

    @Autowired
    private ProjectMemberRepository projectMemberRepository;

    @Autowired
    private ActivityHistoryService activityHistoryService;
    @Autowired
    private TeamService teamService;
    @Autowired
    private UserService userService;
    @Autowired
    private UserRepository userRepository;

    public List<Project> getAllProjects(String teamId) {
        if (teamId == null || teamId.isBlank()) {
            throw new CustomException("No team provided",400);
        }

        return projectRepository.findByTeam(teamService.getTeamById(teamId));
    }

    public Project createProject(String name, String teamId) {
        if (name == null || teamId == null) {
            throw new CustomException("Invalid project data",400);
        }

        String currentUserEmail = userService.getCurrentUserEmail();


        Project newProject = new Project();
        Team team =  teamService.getTeamById(teamId);
        newProject.setName(name);
        newProject.setTeam(team);
        newProject.setCreatedAt(new Date());

        Project savedProject = projectRepository.save(newProject);

        activityHistoryService.logActivity(currentUserEmail,
                "Created new project: " + name + " in team " + teamId
        );

        return savedProject;
    }

    public Project getProjectById(String projectid){
        return projectRepository.findById(projectid).orElseThrow(()-> new CustomException("No Project Exists",400));
    }

    public void addMembersToProject(String projectId, String email) {
        if (projectId == null || email == null || email.isBlank()) {
            throw new CustomException("Invalid member data",400);
        }

        Optional<ProjectMember> existingMember = projectMemberRepository
                .findByProjectAndUser(this.getProjectById(projectId), userService.getUserByEmail(email));

        if (existingMember.isPresent()) {
            throw new CustomException("User already in project",400);
        }

        ProjectMember newMember = new ProjectMember();
        newMember.setProject(projectRepository.findById(projectId).orElseThrow(()->new CustomException("Project not found",400)));
        newMember.setUser(userRepository.findByEmail(email).orElseThrow(()->new CustomException("User not found",400)));

        projectMemberRepository.save(newMember);

        activityHistoryService.logActivity(email,
                "Added user " + email + " to project " + projectId
        );
    }

    public void deleteProject(String projectId) {
        if (projectId == null) {
            throw new CustomException("Project ID cannot be null",400);
        }

        String currentUserEmail = userService.getCurrentUserEmail();

        Optional<Project> project = projectRepository.findById(projectId);

        if (project.isEmpty()) {
            throw new CustomException("Project not found",400);
        }

        projectRepository.deleteById(projectId);
        activityHistoryService.logActivity(currentUserEmail,
                "Deleted project: " + project.get().getName()
        );
    }
}

File: backend/dotnet/Services/ProjectService.cs
using dotnet.Data;
using dotnet.exceptions;
using dotnet.Interfaces;
using dotnet.Models;

namespace dotnet.Services;

public class ProjectService : IProjectService
{
    
    private readonly DataContext _context;
   private readonly TeamService _teamService;
   private readonly UserService _userService;
   private readonly ActivityHistoryService _activityService;

   public ProjectService(DataContext context, TeamService teamService, 
       UserService userService, ActivityHistoryService activityService)
   {
       _context = context;
       _teamService = teamService;
       _userService = userService;
       _activityService = activityService;
   }
   
   public IEnumerable<Project> GetAllProjects(string teamId)
   {
       if (string.IsNullOrWhiteSpace(teamId))
           throw new CustomException("No team provided", 400);

       var team = _teamService.getTeamById(teamId);
       return _context.Projects.Where(p => p.Team == team).ToList();
   }

   public void CreateProject(string name, string teamId)
   {
       if (string.IsNullOrWhiteSpace(name) || string.IsNullOrWhiteSpace(teamId))
           throw new CustomException("Invalid project data", 400);

       var currentUserEmail = _userService.GetCurrentUserEmail();
       var team = _teamService.getTeamById(teamId);
       
       var newProject = new Project
       {
           Name = name,
           Team = team,
           CreatedAt = DateTime.Now
       };

       _context.Projects.Add(newProject);
       _context.SaveChanges();

       _activityService.LogActivity(currentUserEmail, 
           $"Created new project: {name} in team {teamId}");

   }

   public void AddMemberToProject(string projectId, string email)
   {
       if (string.IsNullOrWhiteSpace(projectId) || string.IsNullOrWhiteSpace(email))
           throw new CustomException("Invalid member data", 400);

       var project = GetProjectById(projectId);
       var user = _userService.GetUserByEmail(email);

       var existingMember = _context.ProjectMembers
           .FirstOrDefault(pm => pm.Project == project && pm.User == user);

       if (existingMember != null)
           throw new CustomException("User already in project", 400);

       var newMember = new ProjectMember
       {
           Project = project,
           User = user
       };

       _context.ProjectMembers.Add(newMember);
       _context.SaveChanges();

       _activityService.LogActivity(email, 
           $"Added user {email} to project {projectId}");
   }

   public Project GetProjectById(string projectId)
   {
       return _context.Projects.FirstOrDefault(p => p.Id == projectId) ??
           throw new CustomException("No Project Exists", 400);
   }

   public void DeleteProject(string projectId)
   {
       if (string.IsNullOrWhiteSpace(projectId))
           throw new CustomException("Project ID cannot be null", 400);

       var currentUserEmail = _userService.GetCurrentUserEmail();
       var project = GetProjectById(projectId);

       _context.Projects.Remove(project);
       _context.SaveChanges();

       _activityService.LogActivity(currentUserEmail,
           $"Deleted project: {project.Name}");
   }
}